src/bestcrypt_ve_fmt_plug.c

```diff
@@ -0,0 +1,456 @@
+	p = strtokm(NULL, "$");
+	for (i = 0; i < 24; i++)
+		cs.salt[i] = (atoi16[ARCH_INDEX(p[2*i])] << 4) | atoi16[ARCH_INDEX(p[2*i+1])];
+	p = strtokm(NULL, "$");
+	for (i = 0; i < 0x60; i++)
+		cs.encrypted_data[i] = (atoi16[ARCH_INDEX(p[2*i])] << 4) | atoi16[ARCH_INDEX(p[2*i+1])];
+
+	MEM_FREE(keeptr);
+
+	return (void *)&cs;
+}
+
+static void set_salt(void *salt)
+{
+	cur_salt = (struct custom_salt *)salt;
+}
+
+static void bcve_set_key(char *key, int index)
+{
+	strnzcpy(saved_key[index], key, sizeof(saved_key[index]));
+}
+
+static char *get_key(int index)
+{
+	return saved_key[index];
+}
+
+static int crypt_all(int *pcount, struct db_salt *salt)
+{
+	const int count = *pcount;
+	int index;
+	int failed = 0;
+
+	memset(cracked, 0, sizeof(cracked[0])*cracked_count);
+
+	yescrypt_params_t params = { .N = 0x8000, .r = 16, .p = 1 };
+
+#ifdef _OPENMP
+#pragma omp parallel for
+#endif
+	for (index = 0; index < count; index++) {
+		unsigned char kdf_out[32];
+		/* BestCrypt uses CBC mode with a null IV */
+		unsigned char iv[16] = {0};
+		unsigned char out[0x60];
+		unsigned char intermediate_keys[64];
+
+		SHA256_CTX ctx;
+		unsigned char sha256_hash[32];
+
+#ifdef _OPENMP
+		int t = omp_get_thread_num();
+		if (t >= max_threads) {
+			failed = -1;
+			continue;
+		}
+#else
+		const int t = 0;
+#endif
+		yescrypt_kdf(NULL, &local[t],
+		             (const uint8_t *)saved_key[index],
+		             strlen(saved_key[index]),
+		             (const uint8_t *)cur_salt->salt,
+		             0x18,
+		             &params,
+		             kdf_out, 0x20);
+
+		/*
+		   we will now use output of scrypt as key for desired encryption
+		   algorithm in CBC mode
+		*/
+		if (cur_salt->enc_algoID == aesId) {
+			AES_KEY aes_key;
+
+			AES_set_decrypt_key(kdf_out, 256, &aes_key);
+			AES_cbc_encrypt(cur_salt->encrypted_data,
+			                out,
+			                0x60,
+			                &aes_key,
+			                iv,
+			                AES_DECRYPT);
+		} else if(cur_salt->enc_algoID == twofishId) {
+			Twofish_key tkey;
+
+			Twofish_prepare_key(kdf_out, 32, &tkey);
+			Twofish_Decrypt_no_padding(&tkey,
+			                           cur_salt->encrypted_data,
+			                           out,
+			                           0x60,
+			                           iv);
+		} else if(cur_salt->enc_algoID == serpentId) {
+			uint8_t ks[SERPENT_KS];
+
+			serpent_set_key(kdf_out, ks);
+			serpent_cbc_decrypt(cur_salt->encrypted_data,
+			                    out,
+			                    0x60,
+			                    ks,
+			                    iv);
+		} else if(cur_salt->enc_algoID == camelliaId) {
+			CAMELLIA_KEY ck;
+
+			Camellia_set_key(kdf_out, 256, &ck);
+			Camellia_cbc_encrypt(cur_salt->encrypted_data,
+			                     out,
+			                     0x60,
+			                     &ck,
+			                     iv,
+			                     CAMELLIA_DECRYPT);
+		} /* else if(cur_salt->enc_algoID == rc6Id) {
+			TODO
+		}
+		*/
```

Comment:
Can also use `sizeof` for the 32, 256, and 0x60 above.

Reply:
This is done in https://github.com/openwall/john/pull/4651/commits/468f1e80cf1a73b17e2e524d55eeb97a780b4e61 but some values are still remaining (as 256 or so). I could not find appropriate already defined constants within core.